File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

string checkCardType(string nr);
bool luhnsalg(long creditCardNumber);

int main(void)
{
    string output = "INVALID\n";
    long number = get_long("Number: ");
    char cardNr[20];
    sprintf(cardNr, "%ld", number);

    output = checkCardType(cardNr);

    (luhnsalg(number)) ? printf("%s", output) : printf("INVALID\n");
}

string checkCardType(string nr)
{
    // Check AMEX 34, 37
    if ((nr[0] == '3' && ((nr[1] == '4') || (nr[1] == '7'))) && strlen(nr) == 15)
    {
        return "AMEX\n";
    }

    // Check MASTERCARD 51, 52, 53, 54, or 55
    if (nr[0] == '5' && ((nr[1] - '0') >= 1 && (nr[1] - '0') <= 5) && strlen(nr) == 16)
    {
        return "MASTERCARD\n";
    }

    // Check VISA 4
    if (nr[0] == '4' && (strlen(nr) == 13 || strlen(nr) == 16))
    {
        return "VISA\n";
    }
    return "INVALID\n";
}

bool luhnsalg(long creditCardNumber)
{
    int digits[16]; // Assuming maximum of 16 digits for a credit card number
    int len = 0;

    // Extract digits from the credit card number and store in the array
    while (creditCardNumber > 0)
    {
        digits[len] = creditCardNumber % 10;
        creditCardNumber /= 10;
        len++;
    }

    int sum = 0;
    bool alternate = false;

    for (int i = 0; i < len; i++)
    {
        int digit = digits[i];

        if (alternate)
        {
            digit *= 2;
            if (digit > 9)
            {
                digit -= 9;
            }
        }

        sum += digit;
        alternate = !alternate;
    }

    return (sum % 10 == 0);
}

File: week6/sentimental-credit/credit.py
# TODO
#  MASTERCARD\n  INVALID\n


def main():
    card_number = input("Number: ")

    if len(card_number) not in [13, 15, 16]:
        print("INVALID")

    if luhn(card_number):
        if card_number[:1] == "4":
            print("VISA")
            return
        elif card_number[:2] == "34" or card_number[:2] == "37":
            print("AMEX")
            return
        elif int(card_number[:2]) >= 51 and int(card_number[:2]) <= 55:
            print("MASTERCARD")
            return

    print("INVALID")
    return


def luhn(card_number):
    digits = [int(digit) for digit in reversed(str(card_number))]
    doubled_digits = [
        2 * digit if index % 2 == 1 else digit for index, digit in enumerate(digits)
    ]
    total = sum([(digit // 10 + digit % 10) for digit in doubled_digits])
    return total % 10 == 0


main()
